pkg/controller/queuejob/metrics.go

```diff
@@ -0,0 +1,123 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queuejob
+
+import (
+	arbv1 "github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/apis/controller/v1beta1"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+	"reflect"
+	"runtime"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"time"
+)
+
+var (
+	allocatableCapacityCpu = prometheus.NewGauge(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "allocatable_capacity_cpu",
+		Help:      "Allocatable CPU Capacity (in millicores)",
+	})
+	allocatableCapacityMemory = prometheus.NewGauge(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "allocatable_capacity_memory",
+		Help:      "Allocatable Memory Capacity",
+	})
+	allocatableCapacityGpu = prometheus.NewGauge(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "allocatable_capacity_gpu",
+		Help:      "Allocatable GPU Capacity",
+	})
+	appWrappersCount = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "appwrappers_count",
+		Help:      "AppWrappers count per state",
+	}, []string{"state"})
+)
+
+func init() {
+	klog.V(10).Infof("Registering metrics")
+	metrics.Registry.MustRegister(
+		allocatableCapacityCpu,
+		allocatableCapacityMemory,
+		allocatableCapacityGpu,
+		appWrappersCount,
+	)
+}
+
+func updateMetricsLoop(controller *XController, stopCh <-chan struct{}) {
+	updateMetricsLoopGeneric(controller, stopCh, time.Minute*1, updateAllocatableCapacity)
+	updateMetricsLoopGeneric(controller, stopCh, time.Second*5, updateQueue)
+}
+
+func getFuncName(f interface{}) string {
+	return runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
+}
+
+func updateMetricsLoopGeneric(controller *XController, stopCh <-chan struct{}, d time.Duration, updateFunc func(xController *XController)) {
+	ticker := time.NewTicker(d)
+	go func() {
+		updateFunc(controller)
+		for {
+			select {
+			case <-ticker.C:
+				if klog.V(10).Enabled() {
+					klog.Infof("[updateMetricsLoopGeneric] Update metrics loop tick: %v", getFuncName(updateFunc))
+				}
+				updateFunc(controller)
+			case <-stopCh:
+				if klog.V(10).Enabled() {
+					klog.Infof("[updateMetricsLoopGeneric] Exiting update metrics loop: %v", getFuncName(updateFunc))
+				}
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func updateAllocatableCapacity(controller *XController) {
+	res := controller.GetAllocatableCapacity()
+	allocatableCapacityCpu.Set(res.MilliCPU)
+	allocatableCapacityMemory.Set(res.Memory)
+	allocatableCapacityGpu.Set(float64(res.GPU))
+}
+
+func updateQueue(controller *XController) {
+	awList, err := controller.appWrapperLister.List(labels.Everything())
+
+	if err != nil {
+		klog.Errorf("[updateQueue] Unable to obtain the list of AppWrappers: %+v", err)
```

Comment:
In case of error, should we update the counters to 0 or leave them as is?

Reply:
I think we should leave them as is and retry to update with new value.

Reply:
This update is done periodically. So instead of retrying, we can simply wait for the next update on the next tick.
I dug deeper into the implementation of `controller.appWrapperLister.List(labels.Everything())`, and it seems that since `labels.Everything().Empty()` returns true, there is no way that `controller.appWrapperLister.List()` will return an error. So the discussion seems to be purely theoretical.